src/OpenSage.Game/Graphics/Cameras/RtsCameraController.cs

```diff
@@ -160,31 +167,66 @@ private static float GetKeyMovement(in CameraInputState inputState, Key positive
             return 0;
         }
 
-        void ICameraController.UpdateCamera(Camera camera, in CameraInputState inputState, in TimeInterval gameTime)
+        void ICameraController.UpdateCameraInput(Camera camera, in CameraInputState inputState, in TimeInterval gameTime)
         {
             if (inputState.LeftMouseDown && inputState.PressedKeys.Contains(Key.AltLeft) || inputState.MiddleMouseDown)
             {
                 RotateCamera(inputState.DeltaX, inputState.DeltaY);
+                _panDirection = CameraPanDirection.None;
             }
             else
             {
                 // tested in Zero Hour - rotation always takes precedence, and all panning is halted when rotating.
-                float forwards, right;
+                float up, right;
                 if (inputState.RightMouseDown)
                 {
-                    forwards = -inputState.DeltaY;
-                    right = inputState.DeltaX;
+                    up = -MousePanSpeed * inputState.DeltaY;
+                    right = MousePanSpeed * inputState.DeltaX;
+                }
+                else
+                {
+                    OpenSage.Mathematics.Rectangle bounds = _panel.Frame;
+
+                    if (inputState.LastX < PanBorderWith) right = -1;
+                    else if (inputState.LastX > bounds.Width - PanBorderWith) right = 1;
+                    else right = GetKeyMovement(inputState, Key.Right, Key.Left);
+
+                    if (inputState.LastY < PanBorderWith) up = 1;
+                    else if (inputState.LastY > bounds.Height - PanBorderWith) up = -1;
+                    else up = GetKeyMovement(inputState, Key.Up, Key.Down);
+                }
+                if (up != 0 || right != 0)
```

Comment:
Couldn't you check if the absolute value of abs(up)/abs(down) is greater than the threshold? Floats are likely to not be exactly 0 in many cases

Reply:
Actually values will be 0, when camera isn't moving by user. But you are right, that's the bad way to detect camera movement.